tower-http/src/trace/body.rs

```diff
@@ -77,26 +77,30 @@ where
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
-        let this = self.project();
+        let mut this = self.project();
         let _guard = this.span.enter();
         let result = ready!(this.inner.poll_trailers(cx));
 
         let latency = this.start.elapsed();
 
-        if let Some((classify_eos, mut on_failure)) =
-            this.classify_eos.take().zip(this.on_failure.take())
-        {
-            match &result {
-                Ok(trailers) => {
+        match &result {
+            Ok(trailers) => {
+                if let (Some(classify_eos), Some(on_failure)) =
+                    (this.classify_eos.take(), &mut this.on_failure)
+                {
                     if let Err(failure_class) = classify_eos.classify_eos(trailers.as_ref()) {
                         on_failure.on_failure(failure_class, latency, this.span);
                     }
+                }
 
-                    if let Some((on_eos, stream_start)) = this.on_eos.take() {
-                        on_eos.on_eos(trailers.as_ref(), stream_start.elapsed(), this.span);
-                    }
+                if let Some((on_eos, stream_start)) = this.on_eos.take() {
+                    on_eos.on_eos(trailers.as_ref(), stream_start.elapsed(), this.span);
                 }
-                Err(err) => {
+            }
+            Err(err) => {
+                if let (Some(classify_eos), Some(on_failure)) =
+                    (this.classify_eos.take(), &mut this.on_failure)
+                {
```

Comment:
Note to self: Should probably revert the changes to this file as they don't actually change anything.